Lion search in image set

iezer opened the pull request:
@bantic  For you, CC @redyaffle 

Embedding the lion search below an image set so that users can search. We'll want to tweak the UI for this but the basic structure is there. I did a bunch of refactoring so that I could reuse code from cv-results as much as possible. Such as moving some actions up to the image-set route so that they can get called from either the cv-result or lion-search template.

Here's what it looks like
![](http://g.recordit.co/05fq6VMF3V.gif)

iezer committed:
refactor cv-results page into lg-lion-associator component which can be re-used.

iezer committed:
refactor lion search to make lions selectable and have a single view button

iezer committed:
refactor cv-results to use lg-selectable component

iezer committed:
embedded lion search under image-set page with option to associate image-set with lion

bantic commented on app/routes/image-set.js:
```diff
@@ -15,5 +15,32 @@ export default Ember.Route.extend( OrganizationRouteMixin, RequireUserMixin, {
+
+      imageSet.save().then(function() {
+        route.transitionTo('lion', lion);
+      });
```
what happens if there is an error?

bantic replied:
does the UI give any feedback during the save?

bantic commented on app/routes/image-set.js:
```diff
@@ -15,5 +15,32 @@ export default Ember.Route.extend( OrganizationRouteMixin, RequireUserMixin, {
+
+      lion.save().then(function(lion){
+        route.transitionTo('lion', lion);
+      });
```
same Q as above about an error — does the app just hang at that point?

bantic replied:
same Q as above about saving — does the UI give any feedback that the lion is saving?

bantic commented on app/routes/image-set/lion-search.js:
```diff
@@ -0,0 +1,23 @@
+
+export default Ember.Route.extend(OrganizationRouteMixin, {
+  model: function() {
+    return [];
```
I was wondering for a while why there is an empty array here until I realized the that the controller will populate an array for the `model` property in its `displayResults` action — right? A comment or explanatory variable name would help.

And this is a cognitive mismatch with the place where we set model to `null` in `resetController`, below. It should also set to an empty array if that is the `reset`/default state for this route.

bantic commented on app/templates/image-set/lion-search.hbs:
```diff
@@ -0,0 +1,24 @@
+
+    {{#if model}}
+      <h3>Results</h3>
+      {{#each model as |lion|}}
```
nice use of block params

bantic commented on app/templates/image-set/lion-search.hbs:
```diff
@@ -0,0 +1,24 @@
+      {{#each model as |lion|}}
+        {{#lg-selectable model=lion activeModel=activeLion action='selectLion' classNames="row"}}
+          {{lg-lion-summary lion=lion}}
+          {{/lg-selectable}}
```
indentation

bantic commented on tests/acceptance/image-set/lion-search-test.js:
```diff
@@ -0,0 +1,75 @@
+  signInAndVisit('/image-set/25');
+
+  andThen(function(){
+    click('.lion-search-button');
```
this doesn't have to be in an `andThen`. ember's async test helpers will correctly join up the `signInAndVisit` with the `click`.

bantic commented on tests/acceptance/image-set/lion-search-test.js:
```diff
@@ -0,0 +1,75 @@
+    json.lion_id = 2;
+
+    return this.success(json);
+  });
```
putting this inline like this helps to make it clear when it will be used. But, in case it's not clear, this `stubRequest` will happen before any of the async stuff happens. In other words, this request will be stubbed even for the earlier `visit` and `click`s. Just FYI

bantic commented on tests/acceptance/image-set/lion-search-test.js:
```diff
@@ -0,0 +1,75 @@
+  // click search
+  // select lion
+  // make sure it's selected
+  // click assoc
```
here by accident?

bantic commented:
Cool, looks pretty good, thanks. I'd still like to see UI feedback that a request is inflight or errored. You should be able to use `{{#if imageSet.isSaving}}` but will probalby need another property to do it for the `createLion` action.

What happens if you have a network conditioner set to simulate slow 3g speeds and accidentally double-click on the 'create lion' button?

iezer committed:
Display Errors and Add Front end validations
 - disable UI options where not applicable.
 - adapter creates DS.Errors object.
 - UI validations in image set page
 - Image Set must have at least 1 image (disable delete button if only 1)
 - only public images can become Main Image
 - Main Image cannot flip to Private
 - Save Image set on any image property change using kg-image-controller component
 - Show errors and saving in progress at top of window (good way to test
 this is remove disabled flag from makeMainImage button and try to make
 a private image Main. Back-end will reject)
 - Errors associating image_set with lion or creating new lion displayed
 - tweak list of possible image types to match back-end

iezer committed:
rollback imageSet on failed lion association
 - acceptance tests for model errors.

iezer commented:
@bantic Addressed a bunch of your comments and created a component for showing "is saving" and messages in a template for a given model. 

Also added some logic to the image set page for the following
- save the image set automatically if any image properties change
- disable things we don't want the user to do like
  - making a private image the main
  - making the main image private
  - deleting the last image so that there are no images in an image set. 

Please have another look and then I'd like to merge this.

![](http://g.recordit.co/Xl8EBBPz9f.gif)

iezer committed:
show active image set thumbnail in main image-sets page

bantic commented on app/components/lg-image-gallery.js:
```diff
@@ -9,6 +9,13 @@ export default Ember.Component.extend({
+    saveImageSet: function() {
+      var imageSet = this.get('imageSet');
+      if (imageSet.get('id')) {
+        imageSet.save();
```
does this also need to have its 'saving' and 'error' states handled?

iezer replied:
in progress and error messages will show up in the template while this is saving.

bantic commented on app/routes/image-set.js:
```diff
@@ -15,5 +15,45 @@ export default Ember.Route.extend( OrganizationRouteMixin, RequireUserMixin, {
+      imageSet.save().then(function() {
+        route.transitionTo('lion', lion);
+      }, function(){
+        imageSet.setProperties({
```
probably don't want to just blindly throw away errors here — they might be legitimate API or network failures?

bantic commented on app/routes/image-set.js:
```diff
@@ -15,5 +15,45 @@ export default Ember.Route.extend( OrganizationRouteMixin, RequireUserMixin, {
+      lion.save().then(function(lion){
+        route.transitionTo('lion', lion);
+      }, function(){
+        //error shown in template
```
technically you probably want to ensure that the error is `instanceof DS.InvalidError` -- the error could be something more generic like a syntax error or network failure, in which case you'd want to `return` the error and let somethign else handle it

iezer committed:
error handling
